SF-3191 Improve latest draft retrieval when webhook fails

kylebuss opened the pull request:
Marked as draft as still has work to do. This is a very rough first draft and I wanted to get some feedback before fully implementing this. Mostly on architecture and design for a Helper class like this.

Raymond and I discussed the solution and determined that we would likely need a parser for creating the book list of the latest draft. While SF currently creates a list of individual books (and we could simply `.Split(";")`), Serval accepts a variety of [book ranges](https://github.com/sillsdev/serval/wiki/Filtering-Paratext-Project-Data-with-a-Scripture-Range). I've implemented logic that should work with all acceptable ranges except chapter specific. This is for in the event we would change how we send data to Serval.

We did look into using SIL.Machine's `ScriptureRangeParser` but it's [implementation](https://github.com/sillsdev/machine/blob/master/src/SIL.Machine/Scripture/ScriptureRangeParser.cs) does not appear to meet our needs.

RaymondLuong3 reviewed (commented):
I really would rather be using a library. What is wrong with ScriptureRangeParser.ToChapters() in SIL.Machine that makes it not work for us?

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-:-OMhAg05B0mO_y6mJMI5:bm99jmi)* status: all files reviewed, 6 unresolved discussions (waiting on @kylebuss and @pmachapman)

___
*[`src/SIL.XForge.Scripture/Helpers/ScriptureRangeParser.cs` line 22 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-OMhAQk4BBqO-ZEAmE2L:-OMhAQk55zuOvvtDVpS7:b-6a2o94) ([raw file](https://github.com/sillsdev/web-xforge/blob/d7dc83f8235ebb886dc047071da9488d4579de68/src/SIL.XForge.Scripture/Helpers/ScriptureRangeParser.cs#L22)):*
> ```Smalltalk
>         @"^([A-Z\d]{3}|OT|NT)(, ?([A-Z\d]{3}|OT|NT))*$",
>         RegexOptions.Compiled
>     );
> ```

This doesn't look like it is handling when a book is subtracted, like `GEN-DEU;-NUM`



_Code quote:_
```Smalltalk
    private readonly Regex CommaSeparatedBooks = new Regex(
        @"^([A-Z\d]{3}|OT|NT)(, ?([A-Z\d]{3}|OT|NT))*$",
        RegexOptions.Compiled
    );
```
___
*[`src/SIL.XForge.Scripture/Helpers/ScriptureRangeParser.cs` line 98 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-OMh9BzT-oZvrJd3RlQI:-OMh9BzT-oZvrJd3RlQJ:bqf5qti) ([raw file](https://github.com/sillsdev/web-xforge/blob/d7dc83f8235ebb886dc047071da9488d4579de68/src/SIL.XForge.Scripture/Helpers/ScriptureRangeParser.cs#L98)):*
> ```Smalltalk
>                 break;
>             case "NT":
>                 AddAllBooksForTestament(book);
> ```

This can just be collapsed since the logic is the same. Also, book is a misleading name. Perhaps token would be better



_Code quote:_
```Smalltalk
            case "OT":
                AddAllBooksForTestament(book);
                break;
            case "NT":
                AddAllBooksForTestament(book);
```


<!-- Sent from Reviewable.io -->

kylebuss commented:
@Enkidu93 kind of a side question, but would it be difficult to add a retry to the Serval webhook if it receives a failed message or a bad response code?





<!-- Sent from Reviewable.io -->

Enkidu93 commented:
> @Enkidu93 kind of a side question, but would it be difficult to add a retry to the Serval webhook if it receives a failed message or a bad response code?

It should already. It retries up to 20 times over about 5 hours - or at least it should be!

pmachapman reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3132)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @github-advanced-security[bot] and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 285 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-OMmfyKTDUOlnQCJzQ2t:-OMmfyKTDUOlnQCJzQ2u:byng7j1) ([raw file](https://github.com/sillsdev/web-xforge/blob/7bd79d1d58a3cc33920cb54b4dad17d45866bb89/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L285)):*
> ```Smalltalk
> 
>         // Ensure that the user has permission
>         await EnsureProjectPermissionAsync(curUserId, sfProjectId, isServalAdmin);
> ```

This returns a project, so you could just use the project returned here instead of getting it below.



_Code quote:_
```Smalltalk
await EnsureProjectPermissionAsync(curUserId, sfProjectId, isServalAdmin);
```
___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 325 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-OMme9KeBG32eokp7_QW:-OMme9KeBG32eokp7_QX:b-85zfic) ([raw file](https://github.com/sillsdev/web-xforge/blob/7bd79d1d58a3cc33920cb54b4dad17d45866bb89/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L325)):*
> ```Smalltalk
>                             }
>                         });
>                 }
> ```

I like your approach, but I have concerns about this running every time the latest build is retrieved.

Given a scenario when a user has a book with with one 2 chapters (say Ezra 2 and 3), generates a draft, and then afterwards adds the remaining chapters, this will check for drafts every single time this method is called.

I wonder if we should take your logic here, and use it to replace this block in `PreTranslationService.UpdatePreTranslationStatusAsync()`:
```cs
        Dictionary<int, HashSet<int>> bookChapters = [];
        foreach (
            Pretranslation preTranslation in await translationEnginesClient.GetAllPretranslationsAsync(
                translationEngineId,
                corpusId,
                textId: null,
                cancellationToken
            )
        )
        {
```
as your check is much more efficient, and perhaps just record the `buildId` as a "lastPretranslatedBuildId" or similar in `sf_project_secrets`, and only update if the `buildId` we retrieve here is different to the last one recorded?



_Code quote:_
```Smalltalk
                SFProject project = await projectService.GetProjectAsync(sfProjectId);

                Dictionary<string, List<int>> scriptureRanges = ScriptureRangeParser.GetChapters(
                    project.TranslateConfig.DraftConfig.LastSelectedTranslationScriptureRange
                );

                string[] scriptureRangeIds = [.. scriptureRanges.Keys];

                foreach (string bookId in Canon.AllBookIds.Where(bookId => scriptureRangeIds.Contains(bookId)))
                {
                    int bookNum = Canon.BookIdToNumber(bookId);

                    project
                        .Texts.Where(text => text.BookNum == bookNum)
                        .ForEach(async info =>
                        {
                            if (info.Chapters.Any(c => !c.HasDraft ?? false))
                            {
                                await RetrievePreTranslationStatusAsync(sfProjectId, cancellationToken);
                                return;
                            }
                        });
                }
```


<!-- Sent from Reviewable.io -->

kylebuss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3132)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @github-advanced-security[bot], @pmachapman, and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 285 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-OMmfyKTDUOlnQCJzQ2t:-OMmiTit0zrGvuRIOpwz:b-jz4yz7) ([raw file](https://github.com/sillsdev/web-xforge/blob/7bd79d1d58a3cc33920cb54b4dad17d45866bb89/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L285)):*
<details><summary><i>Previously, pmachapman (Peter Chapman) wrote…</i></summary><blockquote>

This returns a project, so you could just use the project returned here instead of getting it below.
</blockquote></details>

Cool. I will look into this!


<!-- Sent from Reviewable.io -->

kylebuss commented:
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 325 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-OMme9KeBG32eokp7_QW:-OMrBM252cX2LhSIaSwc:bcjtbxu) ([raw file](https://github.com/sillsdev/web-xforge/blob/7bd79d1d58a3cc33920cb54b4dad17d45866bb89/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L325)):*
<details><summary><i>Previously, pmachapman (Peter Chapman) wrote…</i></summary><blockquote>

I like your approach, but I have concerns about this running every time the latest build is retrieved.

Given a scenario when a user has a book with with one 2 chapters (say Ezra 2 and 3), generates a draft, and then afterwards adds the remaining chapters, this will check for drafts every single time this method is called.

I wonder if we should take your logic here, and use it to replace this block in `PreTranslationService.UpdatePreTranslationStatusAsync()`:
```cs
        Dictionary<int, HashSet<int>> bookChapters = [];
        foreach (
            Pretranslation preTranslation in await translationEnginesClient.GetAllPretranslationsAsync(
                translationEngineId,
                corpusId,
                textId: null,
                cancellationToken
            )
        )
        {
```
as your check is much more efficient, and perhaps just record the `buildId` as a "lastPretranslatedBuildId" or similar in `sf_project_secrets`, and only update if the `buildId` we retrieve here is different to the last one recorded?
</blockquote></details>

I have updated the following from our discussion: 
1. Use `EnsureProjectPermissionsAsync` to get the project.
2. Use the scripture ranges from Serval rather than `LastSelectedTranslationScriptureRange`. 
3. Use a Linq query to determine if any chapters `HasDraft` status is null or false when it should be true.

I have not gotten the chance to setup projects to test the scenario if a target/source is missing chapters and will update you once I've tested this.


<!-- Sent from Reviewable.io -->

kylebuss commented:
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 325 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-OMme9KeBG32eokp7_QW:-ON-rODr70BLGQwmukfn:betb90w) ([raw file](https://github.com/sillsdev/web-xforge/blob/7bd79d1d58a3cc33920cb54b4dad17d45866bb89/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L325)):*
<details><summary><i>Previously, kylebuss (Kyle Buss) wrote…</i></summary><blockquote>

I have updated the following from our discussion: 
1. Use `EnsureProjectPermissionsAsync` to get the project.
2. Use the scripture ranges from Serval rather than `LastSelectedTranslationScriptureRange`. 
3. Use a Linq query to determine if any chapters `HasDraft` status is null or false when it should be true.

I have not gotten the chance to setup projects to test the scenario if a target/source is missing chapters and will update you once I've tested this.
</blockquote></details>

Reverted this back to draft status. Need to review existing behavior as testing the target/source missing chapters I observed an interesting behavior in the UI and just want to confirm if this change caused the behavior or not.


<!-- Sent from Reviewable.io -->

kylebuss reviewed (commented):
@pmachapman @RaymondLuong3 This is ready for further review. I've verified that the behavior I was seeing matches what happens on master. 

The latest changes should handle both scenarios if the target or source is missing chapters. I did not investigate it but on both master and this branch if the book was empty (Titus) the draft only added the first chapter and not all the chapters.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-:-ON1XnP32fRdgafmB2Pd:bymdkl9)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @github-advanced-security[bot], @pmachapman, and @RaymondLuong3)



<!-- Sent from Reviewable.io -->

pmachapman reviewed (changes requested):
> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3132)* status: 4 of 5 files reviewed, 5 unresolved discussions (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 321 at r4](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-ONHDz6x16R4WrwhX7wZ:-ONHDz6x16R4WrwhX7w_:b-xn0pzt) ([raw file](https://github.com/sillsdev/web-xforge/blob/351f310e48d07f4332f511e876c8c960df540517/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L321)):*
> ```Smalltalk
>                             .ToDictionary(kvp => kvp.Key, kvp => kvp.Value, StringComparer.OrdinalIgnoreCase);
>                     }
>                 }
> ```

You could probably simplify this to:
```cs
                Dictionary<string, List<int>> scriptureRanges = (translationBuild.Pretranslate ?? [])
                    .SelectMany(pc => pc.SourceFilters ?? [])
                    .SelectMany(pcf => ScriptureRangeParser.GetChapters(pcf.ScriptureRange))
                    .ToDictionary(kvp => kvp.Key, kvp => kvp.Value, StringComparer.OrdinalIgnoreCase);
```

However I notice that the values are not merged if scripture ranges share the same book. Serval supports (and we will eventually implement) translating some chapters with one source and other chapters with another. Also, we need to use the correct versification for the project when calculating the project chapters, and as noted in the next comment, ScriptureRangeParser.GetChapters() returns an empty list if all of the chapters in a book are selected (at least in my tests?).

We should instead handle it something like this:

```cs
 // TODO: Get the correct versification for the project from ParatextData's ProjectSettings.Versification
ScrVers versification = ScrVers.English;
ScriptureRangeParser scriptureRangeParser = new ScriptureRangeParser(versification);

foreach (PretranslateCorpus ptc in translationBuild.Pretranslate ?? [])
{
    // We are using the TranslationBuild.Pretranslate.SourceFilters.ScriptureRange to find the
    // books selected for drafting. Some projects may have used the now obsolete field
    // TranslationBuild.Pretranslate.ScriptureRange and will not get checked for webhook failures.
    foreach (ParallelCorpusFilter source in ptc.SourceFilters ?? [])
    {
        foreach ((string book, List<int> bookChapters) in scriptureRangeParser.GetChapters(source.ScriptureRange))
        {
            // Ensure that if chapters is blank, it contains every chapter in the book
            List<int> chapters = bookChapters;
            if (chapters.Count == 0)
            {
                chapters = [.. Enumerable.Range(1, versification.GetLastChapter(Canon.BookIdToNumber(book)))];
            }

            // Set or merge the list of chapters
            if (!scriptureRanges.TryGetValue(book, out List<int> existingChapters))
            {
                scriptureRanges[book] = chapters;
            }
            else
            {
                // Merge new chapters into existing list, avoiding duplicates
                foreach (int chapter in chapters.Where(chapter => !existingChapters.Contains(chapter)))
                {
                    existingChapters.Add(chapter);
                }
            }
        }
    }
}
```



_Code quote:_
```Smalltalk
                IList<PretranslateCorpus> pretranslateCorpus = translationBuild.Pretranslate;

                foreach (PretranslateCorpus ptc in pretranslateCorpus)
                {
                    // We are using the TranslationBuild.Pretranslate.SourceFilters.ScriptureRange to find the
                    // books selected for drafting. Some projects may have used the now obsolete field
                    // TranslationBuild.Pretranslate.ScriptureRange and will not get checked for webhook failures.
                    foreach (ParallelCorpusFilter source in ptc.SourceFilters)
                    {
                        Dictionary<string, List<int>>? scriptureRange = ScriptureRangeParser.GetChapters(
                            source.ScriptureRange
                        );

                        scriptureRanges = scriptureRanges
                            .Union(scriptureRange)
                            .ToDictionary(kvp => kvp.Key, kvp => kvp.Value, StringComparer.OrdinalIgnoreCase);
                    }
                }
```
___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 335 at r4](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-ONHQaOb7TLgwtZLB40N:-ONHQaOb7TLgwtZLB40O:b-jyypz) ([raw file](https://github.com/sillsdev/web-xforge/blob/351f310e48d07f4332f511e876c8c960df540517/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L335)):*
> ```Smalltalk
>                             && text.Chapters.Where(chapter => scriptureRanges[book].Contains(chapter.Number))
>                                 .All(c => c.HasDraft ?? false)
>                         );
> ```

You are checking for the opposite of what the variable is called (your logic is saying "Do all chapters have HasDraft set to true).

This instead should look something like:

```cs
                bool hasDraftMissingFromChapters = Canon
                    .AllBookIds.Where(scriptureRanges.ContainsKey)
                    .All(book => project.Texts.Any(text =>
                        text.BookNum == Canon.BookIdToNumber(book)
                        && text.Chapters.Where(chapter => scriptureRanges[book].Contains(chapter.Number))
                            .All(c => !(c.HasDraft ?? false))
                    ));
```

Before fixing this bug, perhaps create a test case that will fail and expose this bug?



_Code quote:_
```Smalltalk
                bool hasDraftMissingFromChapters = Canon
                    .AllBookIds.Where(book => scriptureRangeIds.Contains(book))
                    .All(book =>
                    {
                        int bookNum = Canon.BookIdToNumber(book);
                        return project.Texts.Any(text =>
                            text.BookNum == bookNum
                            && text.Chapters.Where(chapter => scriptureRanges[book].Contains(chapter.Number))
                                .All(c => c.HasDraft ?? false)
                        );
```
___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 341 at r4](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-ONHLp9UCR47PuJ0eV2m:-ONHLp9UCR47PuJ0eV2n:bgskbfm) ([raw file](https://github.com/sillsdev/web-xforge/blob/351f310e48d07f4332f511e876c8c960df540517/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L341)):*
> ```Smalltalk
>                 {
>                     // Chapters HasDraft is missing or false but should be true, retrieve the pre-translation status to update them.
>                     await RetrievePreTranslationStatusAsync(sfProjectId, cancellationToken);
> ```

I created a draft for 2PE, and every time I viewed the build page, this ran because `ScriptureRangeParser.GetChapters(source.ScriptureRange)` returns an empty list for a book if no subset of chapters are specified, and your check performed the opposite of what it should. (see both code suggestions above)



_Code quote:_
```Smalltalk
await RetrievePreTranslationStatusAsync(sfProjectId, cancellationToken);
```
___
*[`test/SIL.XForge.Scripture.Tests/Services/MachineApiServiceTests.cs` line 1099 at r4](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-ONHSGTj8WDcZNeiMX-5:-ONHSGTj8WDcZNeiMX-6:bsu2ev5) ([raw file](https://github.com/sillsdev/web-xforge/blob/351f310e48d07f4332f511e876c8c960df540517/test/SIL.XForge.Scripture.Tests/Services/MachineApiServiceTests.cs#L1099)):*
> ```Smalltalk
>                                             ScriptureRange = "GEN",
>                                         },
>                                     ],
> ```

This is too simplistic. As a large part of your logic is about handling multiple parallel corpus filters, can you create tests that expand these checks? Perhaps values like:

```cs
                                    SourceFilters =
                                    [
                                        new ParallelCorpusFilter
                                        {
                                            Corpus = { Id = "corpusId1", Url = "https://example.com" },
                                            ScriptureRange = "GEN1-2",
                                        },
                                        new ParallelCorpusFilter
                                        {
                                            Corpus = { Id = "corpusId2", Url = "https://example.com" },
                                            ScriptureRange = "GEN3-4;EXO",
                                        },
                                        new ParallelCorpusFilter
                                        {
                                            Corpus = { Id = "corpusId2", Url = "https://example.com" },
                                            ScriptureRange = "EXO",
                                        },
                                        new ParallelCorpusFilter
                                        {
                                            Corpus = { Id = "corpusId2", Url = "https://example.com" },
                                            ScriptureRange = "LEV",
                                        },
                                    ],
```
(have a think on other edge cases too - for example, I notice that you check the book ids against Canon.AllBookIds - Have ParallelCorpusFilter where the ScriptureRange is empty or bogus values like "INVALID")



_Code quote:_
```Smalltalk
                                    SourceFilters =
                                    [
                                        new ParallelCorpusFilter
                                        {
                                            Corpus = new ResourceLink { Id = "corpusId", Url = "https://example.com" },
                                            ScriptureRange = "GEN",
                                        },
                                    ],
```
___
*[`test/SIL.XForge.Scripture.Tests/Services/MachineApiServiceTests.cs` line 1107 at r4](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-ONHRm-tE_aq_Ly5CfOe:-ONHRm-tE_aq_Ly5CfOf:b-ui8o88) ([raw file](https://github.com/sillsdev/web-xforge/blob/351f310e48d07f4332f511e876c8c960df540517/test/SIL.XForge.Scripture.Tests/Services/MachineApiServiceTests.cs#L1107)):*
> ```Smalltalk
>             );
>         SFProject project = env.Projects.Get(Project01);
>         project.Texts[0].Chapters[1].HasDraft = true;
> ```

This is kind of hacky. Also, the fact that the tests did not pick up the logic error I noted in `hasDraftMissingFromChapters` means that the test is not a good test. Can you implement tests that have all chapters with drafts, so code errors like the logic error in `hasDraftMissingFromChapters` can be picked up?



_Code quote:_
```Smalltalk
        SFProject project = env.Projects.Get(Project01);
        project.Texts[0].Chapters[1].HasDraft = true;
```


<!-- Sent from Reviewable.io -->

pmachapman commented:
@RaymondLuong3 Are you able to review this PR? I have taken it over from Kyle, so can no longer be objective in my review.

pmachapman reviewed (commented):
> Dismissed @github-advanced-security[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3132)* status: 0 of 7 files reviewed, 5 unresolved discussions (waiting on @kylebuss and @RaymondLuong3)



<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (approved):
Thanks for your work on this. I tested this out and it is working as I would expect. Can you fill out the acceptance tests on the issue, or confirm that they are still accurate?

> Reviewed 1 of 5 files at r2, 1 of 4 files at r5, 4 of 5 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-:-OQAmlj06gY-ZTbRzHes:b-p8vye5)* status: all files reviewed, 7 unresolved discussions (waiting on @kylebuss and @pmachapman)

___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 447 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-OQAmLyU0efGWJYMttWy:-OQAmLyU0efGWJYMttWz:b-2bhboz) ([raw file](https://github.com/sillsdev/web-xforge/blob/ae0e7877b3ed6be9a91b53e0fa78b5904b51fa45/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L447)):*
> ```Smalltalk
>                             if (!scriptureRanges.TryGetValue(book, out List<int> existingChapters))
>                             {
>                                 scriptureRanges[book] = chapters;
> ```

Nit: The may be better named scriptureRangesWithDrafts



_Code quote:_
```Smalltalk
scriptureRanges[book] = chapters;
```
___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 465 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-OQAl3CjBwc28nQLjsMo:-OQAl3CjBwc28nQLjsMp:bpmz6cm) ([raw file](https://github.com/sillsdev/web-xforge/blob/ae0e7877b3ed6be9a91b53e0fa78b5904b51fa45/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L465)):*
> ```Smalltalk
>                 // check if any chapters from the scripture range are marked as HasDraft = false or null
>                 bool hasDraftIsFalseOrNullInScriptureRange =
>                     scriptureRangeBooks.Count > 0
> ```

Nit: Maybe this would be better called scriptureRangeBooksWithDraft



_Code quote:_
```Smalltalk
                    scriptureRangeBooks.Count > 0
```


<!-- Sent from Reviewable.io -->

pmachapman reviewed (approved):
> Can you fill out the acceptance tests on the issue, or confirm that they are still accurate?

Done.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-:-OQB6fR74w7Ffi5W7Dc9:b-8fntoe)* status: all files reviewed, 5 unresolved discussions (waiting on @kylebuss and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 447 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-OQAmLyU0efGWJYMttWy:-OQB7NsL8DofUHkr8rJ_:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/ae0e7877b3ed6be9a91b53e0fa78b5904b51fa45/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L447)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Nit: The may be better named scriptureRangesWithDrafts
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 465 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/3132#-OQAl3CjBwc28nQLjsMo:-OQB72S76BaZnBl4SBix:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/ae0e7877b3ed6be9a91b53e0fa78b5904b51fa45/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L465)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Nit: Maybe this would be better called scriptureRangeBooksWithDraft
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (approved):
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3132)* status: all files reviewed, 5 unresolved discussions (waiting on @kylebuss and @pmachapman)



<!-- Sent from Reviewable.io -->
